stan/math/rev/core/vari.hpp

```diff
@@ -184,6 +186,400 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  explicit vari_value(S&& x)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        rows_(x.rows()),
+        cols_(x.cols()),
+        size_(x.size()),
+        adj_(make_adj(adj_mem_, x)),
+        val_(make_val(val_mem_, std::forward<S>(x))) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct an dense Eigen variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
+   *  should be constructed before variables on which they depend
+   *  to insure proper partial derivative propagation.  During
+   *  derivative propagation, the chain() method of each variable
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   * @param stacked If false will put this this vari on the nochain stack so
+   * that its `chain()` method is not called.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  vari_value(S&& x, bool stacked)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        rows_(x.rows()),
+        cols_(x.cols()),
+        size_(x.size()),
+        adj_(make_adj(adj_mem_, x)),
+        val_(make_val(val_mem_, x)) {
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
+
+  /**
+   * Return the number of rows for this class's `val_` member
+   */
+  const Eigen::Index rows() const { return rows_; }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  const Eigen::Index cols() const { return cols_; }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  const Eigen::Index size() const { return size_; }
+
+  virtual void chain() {}
+  /**
+   * Initialize the adjoint for this (dependent) variable to 1.
+   * This operation is applied to the dependent variable before
+   * propagating derivatives, setting the derivative of the
+   * result with respect to itself to be 1.
+   */
+  void init_dependent() { adj_.setOnes(); }
```

Comment:
There's no obvious default for this. I am confused as to what to do here. I think maybe we should just turn off `.grad()` and `.init_dependent()` for these types.

Reply:
Confused by what you mean here. Why wouldn't we set the matrix to ones?

Reply:
By the above I mean to ask why would `.grad()` and `init_dependent()` not make sense?

Reply:
As far as I know, the adjoints in a reverse mode pass always represent gradients of one value. When we do:

```
var a;
var b = a * a;
b.grad()
```

then the adjoint of `b` is `db/db` and the adjoint of `a` is `db/da`.

If we had something like:
```
var_value<Matrix> a;
var_value<Matrix> b = a * a;
b.grad();
```

What do the adjoints of `b` or the adjoints of `a` mean on their own?

Initializing the adjoints of `b` to all 1s is like implicitly defining `b.grad();` as equivalent to `sum(b).grad()`.

I think it's entirely reasonably you'd want to initialize the adjoints of `b` to something and then call `grad(b.vi_)`. That's how  you'd compute `row_vector * Jacobian(f)` products, but then you'd want to initialize the adjoints to some specific values and call grad manually, not work through the `b.grad()` + `init_dependent` interface.

I don't think I did a terribly good job of making myself clear there. @bob-carpenter and @betanalpha might have suggestions here on the interface here.